Code/GraphMol/GenericGroups/GenericGroups.cpp

```diff
@@ -0,0 +1,481 @@
+  // redundant
+  if (!mol.getRingInfo() || !mol.getRingInfo()->isInitialized()) {
+    MolOps::fastFindRings(mol);
+  }
+  auto atomMatcher = [](const Atom &at) -> bool {
+    return !at.getIsAromatic() &&
+           (at.getAtomicNum() == 6 || at.getAtomicNum() == 1);
+  };
+  auto bondMatcher = [extraBondType](const Bond &bnd) -> bool {
+    return (bnd.getBondType() == Bond::BondType::SINGLE ||
+            bnd.getBondType() == extraBondType) &&
+           !bnd.getIsAromatic() && !queryIsBondInRing(&bnd);
+  };
+  auto atLeastMatcher = [extraBondType](const Bond &bnd) -> bool {
+    return bnd.getBondType() == extraBondType;
+  };
+  AtomMatcherFunc atomAtLeast = nullptr;
+  return AllAtomsMatch(mol, atom, ignore, atomMatcher, bondMatcher, atomAtLeast,
+                       atLeastMatcher);
+}
+}  // namespace
+
+bool AlkenylAtomMatcher(const ROMol &mol, const Atom &atom,
+                        boost::dynamic_bitset<> ignore) {
+  return UnsatAlkXAtomMatcher(mol, atom, ignore, Bond::BondType::DOUBLE);
+}
+
+bool AlkynylAtomMatcher(const ROMol &mol, const Atom &atom,
+                        boost::dynamic_bitset<> ignore) {
+  return UnsatAlkXAtomMatcher(mol, atom, ignore, Bond::BondType::TRIPLE);
+}
+
+namespace {
+bool checkAtomRing(const ROMol &mol, const Atom &atom,
+                   const boost::dynamic_bitset<> &ignore,
+                   const std::vector<int> &ring, AtomMatcherFunc matcher,
+                   AtomMatcherFunc atLeastOne) {
+  bool atLeast = atLeastOne == nullptr;
+  for (auto aidx : ring) {
+    if (aidx != static_cast<int>(atom.getIdx()) &&
+        (ignore[aidx] || (matcher && !matcher(*mol.getAtomWithIdx(aidx))))) {
+      return false;
+    }
+    if (!atLeast && atLeastOne(*mol.getAtomWithIdx(aidx))) {
+      atLeast = true;
+    }
+  }
+  return atLeast;
+}
+bool checkBondRing(const ROMol &mol, const std::vector<int> &bring,
+                   BondMatcherFunc matcher, BondMatcherFunc atLeastOne) {
+  bool atLeast = atLeastOne == nullptr;
+  for (auto bidx : bring) {
+    if (matcher && !matcher(*mol.getBondWithIdx(bidx))) {
+      return false;
+    }
+    if (!atLeast && atLeastOne(*mol.getBondWithIdx(bidx))) {
+      atLeast = true;
+    }
+  }
+  return atLeast;
+}
+
+bool FusedRingMatch(const ROMol &mol, const Atom &atom,
+                    boost::dynamic_bitset<> ignore,
+                    AtomMatcherFunc atomMatcher = nullptr,
+                    BondMatcherFunc bondMatcher = nullptr,
+                    AtomMatcherFunc atLeastOneAtomPerRing = nullptr,
+                    BondMatcherFunc atLeastOneBondPerRing = nullptr,
+                    AtomMatcherFunc atLeastOneAtom = nullptr) {
+  PRECONDITION(&atom.getOwningMol() == &mol, "atom not owned by molecule");
+  if (atomMatcher && !atomMatcher(atom)) {
+    return false;
+  }
+
+  if (!mol.getRingInfo() || !mol.getRingInfo()->isInitialized()) {
+    MolOps::findSSSR(mol);
+  }
```

Comment:
I'm seeing this requires SSSR, so I expect this to fail/return a false answer if we use it on a `mol` that had ring info initialized by `fastFindRings()` (as other checks in this review do). Should we think about having a separate check to differentiate ring info provided by `fastFindRings()` and `findSSSR()` ? (not to be added in this PR, of course)

Reply:
I think that anyone who uses fastFindRings should expect what they get.  The docs are pretty clear in C++ land, but not necessarily in python land:

C++
```
//! use a DFS algorithm to identify ring bonds and atoms in a molecule
/*!
  \b NOTE: though the RingInfo structure is populated by this function,
  the only really reliable calls that can be made are to check if
  mol.getRingInfo().numAtomRings(idx) or mol.getRingInfo().numBondRings(idx)
  return values >0
*/
```

Python:

```
Does a non-SSSR ring finding for a molecule.\n\
```

Perhaps the python docs should be updated here as I have seen some confusion in people who prefer fastFindRings because of the "fast" in the title :)

Reply:
I also think, perhaps, we could tag the molecule with the ring finding version used.  This would allow us to throw a warning or make the right ring algorithm.  But this might be solving a 0.5% problem.

Reply:
> I also think, perhaps, we could tag the molecule with the ring finding version used.

That's what I was asking about.

My worries here are that sometimes we run ring finding under the hood. E.g., we import a SMILES string with `sanitize=False`, and depending on what we are calling afterwards, ring info will be initialized using using either `fastFindRings()` or `findSSSR()`, depending on what we do.

If we used `findSSSR()`, then there will be no problem, but if we used `fastFindRings()`, and then try to run something that requires SSSR, I think we might just get wrong answers, since ring info has been initialized, but we did not look for SSSRs. And this would happen silently, since we can't currently tell what we used to find ring info.

Reply:
My two cents on this: if you are using an advanced feature like partial sanitization then it's up to you to make sure you don't shoot yourself in the foot. The RDKit can't reasonably protect you from yourself if you're using advanced user tools.

Reply:
Sorry to come back to this after so long (I've been off for a few days) and continue arguing. Maybe we should move this into a separate discussion?

I agree that we cannot prevent ourselves from doing stupid things with the RDKit. But we are already doing it to some extent: if we try to run some operations (like the queries in this PR) on something that has no ring information, instead of issuing an error message and failing, we run ring finding under the hood (either `fastFindRings()` or `findSSSR()` depending on what we are trying to do, which is what motivates my concern) and still perform the operation.

Reply:
I'm not violently opposed to keeping track of which ring-finding algorithm was used, but it's probably a discussion for a different issue.